src/plugins/kibana_utils/public/history/redirect_when_missing.tsx

```diff
@@ -0,0 +1,83 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import React from 'react';
+import { History } from 'history';
+import { i18n } from '@kbn/i18n';
+
+import { ToastsSetup } from 'kibana/public';
+import { MarkdownSimple, toMountPoint } from '../../../kibana_react/public';
+import { SavedObjectNotFound } from '../errors';
+
+interface Mapping {
+  [key: string]: string;
+}
+
+/**
+ * Creates an error handler that will redirect to a url when a SavedObjectNotFound
+ * error is thrown
+ */
+export function redirectWhenMissing({
+  history,
+  mapping,
+  toastNotifications,
+}: {
+  history: History;
+  /**
+   * a mapping of url's to redirect to based on the saved object that
+   * couldn't be found, or just a string that will be used for all types
+   */
+  mapping: string | Mapping;
+  /**
+   * Toast notifications service to show toasts in error cases.
+   */
+  toastNotifications: ToastsSetup;
+}) {
+  let localMappingObject: Mapping;
+
+  if (typeof mapping === 'string') {
+    localMappingObject = { '*': mapping };
+  } else {
+    localMappingObject = mapping;
+  }
+
+  return (error: SavedObjectNotFound | Error) => {
+    // if this error is not "404", rethrow
+    if (!(error instanceof SavedObjectNotFound)) {
+      throw error;
+    }
+
+    const unknownVisType = error.message.indexOf('Invalid type') === 0;
```

Comment:
Could we move this logic into the visualize consumer (maybe by having a config parameter `mapSavedObjectType` on  the function)? The util should not contain logic specifically for visualizations

Reply:
Good point!
Moved straight into the `editor.js` as a particular case